From 14468defc0f39953e4005ff3edcbd68b14a0d87d Mon Sep 17 00:00:00 2001
From: Jonathan Tinkham <sctincman@gmail.com>
Date: Fri, 1 Apr 2016 14:50:51 -0600
Subject: [PATCH] AsoC: tegra_max98090: honor headphone/mic detect GPIO
 polarity

Set the invert property for the headphone and mic jack depending on the
GPIO polarity in the device-tree.

Signed-off-by: Jonathan Tinkham <sctincman@gmail.com>
---
 sound/soc/tegra/tegra_max98090.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/sound/soc/tegra/tegra_max98090.c b/sound/soc/tegra/tegra_max98090.c
index 902da36..f4df25d 100644
--- a/sound/soc/tegra/tegra_max98090.c
+++ b/sound/soc/tegra/tegra_max98090.c
@@ -42,7 +42,9 @@
 struct tegra_max98090 {
 	struct tegra_asoc_utils_data util_data;
 	int gpio_hp_det;
+	enum of_gpio_flags gpio_hp_det_flags;
 	int gpio_mic_det;
+	enum of_gpio_flags gpio_mic_det_flags;
 };
 
 static int tegra_max98090_asoc_hw_params(struct snd_pcm_substream *substream,
@@ -110,7 +112,6 @@ static struct snd_soc_jack_gpio tegra_max98090_hp_jack_gpio = {
 	.name = "Headphone detection",
 	.report = SND_JACK_HEADPHONE,
 	.debounce_time = 150,
-	.invert = 1,
 };
 
 static struct snd_soc_jack tegra_max98090_mic_jack;
@@ -126,7 +127,6 @@ static struct snd_soc_jack_gpio tegra_max98090_mic_jack_gpio = {
 	.name = "Mic detection",
 	.report = SND_JACK_MICROPHONE,
 	.debounce_time = 150,
-	.invert = 1,
 };
 
 static const struct snd_soc_dapm_widget tegra_max98090_dapm_widgets[] = {
@@ -155,6 +155,8 @@ static int tegra_max98090_asoc_init(struct snd_soc_pcm_runtime *rtd)
 				      ARRAY_SIZE(tegra_max98090_hp_jack_pins));
 
 		tegra_max98090_hp_jack_gpio.gpio = machine->gpio_hp_det;
+		tegra_max98090_hp_jack_gpio.invert =
+			machine->gpio_hp_det_flags & OF_GPIO_ACTIVE_LOW;
 		snd_soc_jack_add_gpios(&tegra_max98090_hp_jack,
 					1,
 					&tegra_max98090_hp_jack_gpio);
@@ -168,6 +170,8 @@ static int tegra_max98090_asoc_init(struct snd_soc_pcm_runtime *rtd)
 				      ARRAY_SIZE(tegra_max98090_mic_jack_pins));
 
 		tegra_max98090_mic_jack_gpio.gpio = machine->gpio_mic_det;
+		tegra_max98090_mic_jack_gpio.invert =
+			machine->gpio_mic_det_flags & OF_GPIO_ACTIVE_LOW;
 		snd_soc_jack_add_gpios(&tegra_max98090_mic_jack,
 				       1,
 				       &tegra_max98090_mic_jack_gpio);
@@ -234,12 +238,13 @@ static int tegra_max98090_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, card);
 	snd_soc_card_set_drvdata(card, machine);
 
-	machine->gpio_hp_det = of_get_named_gpio(np, "nvidia,hp-det-gpios", 0);
+	machine->gpio_hp_det = of_get_named_gpio_flags(
+		np, "nvidia,hp-det-gpios", 0, &machine->gpio_hp_det_flags);
 	if (machine->gpio_hp_det == -EPROBE_DEFER)
 		return -EPROBE_DEFER;
 
-	machine->gpio_mic_det =
-			of_get_named_gpio(np, "nvidia,mic-det-gpios", 0);
+	machine->gpio_mic_det = of_get_named_gpio_flags(
+		np, "nvidia,mic-det-gpios", 0, &machine->gpio_mic_det_flags);
 	if (machine->gpio_mic_det == -EPROBE_DEFER)
 		return -EPROBE_DEFER;
 
-- 
2.8.0

